File: Stationary/matlab/Pulses/dataprocess.m
% data proccessing

close all
clear all
format long
Fs=100e3; %Sample freq

fid=fopen("datos/ruidoR.dat",'rb');
ruidoR=fread(fid,'float');

fid=fopen("datos/ruidoI.dat",'rb');
ruidoI=fread(fid,'float');

ruidoC=complex(ruidoR,ruidoI);  %transmitted data

fid=fopen("datos/dataR.dat",'rb');
dataR=fread(fid,'float');

fid=fopen("datos/dataI.dat",'rb');
dataI=fread(fid,'float');

dataC=complex(dataR,dataI); % received data

lenR=size(ruidoC(:,1));
lenR=lenR(1,1);

lenD=size(dataC(:,1));
lenD=lenD(1,1);

lenDr=size(dataR(:,1));
lenDr=lenDr(1,1);

figure
subplot(2,1,1)
plot(0:1/Fs:lenR/Fs-1/Fs,ruidoR,0:1/Fs:lenR/Fs-1/Fs,ruidoI)
title('Tx real part')
xlabel('time [s]')
subplot(2,1,2)
plot(0:1/Fs:lenDr/Fs-1/Fs,dataR,0:1/Fs:lenDr/Fs-1/Fs,dataI)
title('Rx real part')
xlabel('time [s]')

offset = 371397; % received samples offset

dataC=dataC(offset:lenD);

lenD=size(dataC(:,1));
lenD=lenD(1,1);


F=1000; % Pulse freq
N=Fs/F; % window duration [samples], should be the duration of a pulse  

v=floor(lenD/N); % number of pulses recorded






y=zeros(N,v);

for i=0: v-1% Run cross correlation for v times
    
   

    
   

    x=ruidoC(i*N+1:i*N+N,1); % TX
    y=dataC(i*N+1:i*N+N,1); % RX
    
    rxy=xcorr(x,conj(y)); % Cross correlation of the TX and RX conjugated data
    Ryx=flip(rxy(1:N)); % Flip the correlation result and take the first N samples (Ryx(t) = Rxy(-t)
   %Ryx=flip(rxy);

PA(:,i+1)=Ryx'; %Store the results in an array
end
clear Ryx; 
 
Ryx=sum(PA')/v; % Average the values 

%figure;
%plot(abs(Ryx))

figure;
plot(0:1/Fs:N/Fs-1/Fs,abs(Ryx)); % Plot the estimated impulse response
title('Estimated Impulse Response');
xlabel('time [s]')
grid;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
hold on
R=0.05;%ratio

Pulso = zeros(1,N);

Pulso(1,1:Fs/F*R) = 1; 
att=0.25;

plot(0:1/Fs:N/Fs-1/Fs,Pulso*att);

%QPulso = [zeros(1,Fs/F/4) Pulso]; % 90 degree delay, quadrature signal
%QPulso = QPulso(1,1:N);
%plot(0:1/Fs:N/Fs-1/Fs,QPulso*att);

hold off
%%%%%%%%%%%%%%%%%%%%5

[maximum,pico]=max(abs(Ryx));
pico=pico/Fs

f = fopen ("datos/ImpResp.dat", 'ab');
v = fwrite (f, abs(Ryx),'float');
fclose (f);

File: modules/dataProcess.py
#! /usr/bin/python
import sys
import numpy as np 
import matplotlib.pyplot as plt
from offcalc import offcalc
from SNRcalc import SNRcalc


#%%%%%%%%%5%%%%%%%%%%%%%
def dataProcess(Fs , #Sample freq
	wd = 100,
	path = "/dev/shm/",
	offman = 0,
	plotMode=False):
	ruidoR = np.fromfile(path + "ruidoR.dat",'float32')
	ruidoI = np.fromfile(path + "ruidoI.dat",'float32')


	if len(ruidoR) != len(ruidoI):
		print '!'*40
		print "WARNING! SIZE ERROR OF TX SAMPLES"
		print "I:{}".format(len(ruidoR))
		print "Q:{}".format(len(ruidoI))
		print '!'*40
	
	ruidoC = ruidoR + 1j * ruidoI


	del ruidoR,ruidoI

	dataR = np.fromfile(path + "fdataR.dat",'float32')
	dataI = np.fromfile(path + "fdataI.dat",'float32')


	if len(dataR) != len(dataI):
		print '!'*40
		print "WARNING! SIZE ERROR OF RX SAMPLES"
		print "I:{}".format(len(dataR))
		print "Q:{}".format(len(dataI))
		print '!'*40
	
	dataC = dataR[0:int(2*Fs)] + 1j * dataI[0:int(2*Fs)]
	del dataR,dataI

	#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
	#Calibration processing

	calibrationOffset = 0.6 *Fs  #conversion from time to samples

	# signal calibration
	offset = offcalc(dataC[0:int(Fs*0.6)],Fs) + offman #received samples offset due to hardware & software lag [samples];
	snr = SNRcalc(dataC[0:int(Fs*0.8)],Fs,offset)

	print "SNR: {}".format(snr)

	ruidoC = ruidoC[ int(calibrationOffset) : int(calibrationOffset*Fs) ];
	dataC  = dataC [ offset : int(offset+Fs)  ];


	lenR=len(ruidoC)
	lenD=len(dataC)


	if 1 ==1 :
		N=int( wd*Fs/1000000) # conversion of window duration from miliseconds to samples

		v=int(np.floor(lenD/N)) # number of pulses recorded
		PA = np.zeros((v,N)) + 1j * np.zeros((v,N))


		for i in range(0,v):# Run cross correlation for v times
		    
			x=ruidoC[int(i*N) : int(i*N+N)] #TX
			y=dataC[int(i*N)  : int(i*N+N)] # RX
			
			rxy=np.correlate( x, np.conj(y) , 'full' ) # Cross correlation of the TX and RX conjugated data
			Ryx=np.flip(rxy[0:N],0) # Flip the correlation result and take the first N samples (Ryx(t) = Rxy(-t)
			PA[i] = Ryx

		Ryx = PA.mean(axis=0)
		Ryx = np.abs(Ryx)
	else:
		Ryx = 0

	if plotMode :
		plt.stem(np.abs(np.square(Ryx)))
		plt.show()
	else:
		return Ryx


def main():
	Fs = 38e6 #Sample freq
	wd = 10; # window duration [us] for the correlation purpose
	path = "/dev/shm/"

	dataProcess(Fs,wd,path,plotMode = True)


if __name__ == '__main__':
    main()
